cmd/proxyarpd/main.go

```diff
@@ -31,47 +32,44 @@ func main() {
 		log.Fatalf("invalid IPv4 address: %q", *ipFlag)
 	}
 
+	client, err := arp.NewClient(ifi)
+	if err != nil {
+		log.Fatalf("couldn't create ARP client: %s", err)
+	}
+
 	// Handle ARP requests bound for designated IPv4 address, using proxy ARP
 	// to indicate that the address belongs to this machine
-	proxyARP := func(w arp.ResponseSender, r *arp.Request) {
+	for {
+		pkt, _, err := client.Read()
+		if err != nil {
+			if err == io.EOF {
+				log.Println("EOF")
+				break
+			}
+			log.Fatalf("error processing ARP requests: %s", err)
+		}
+
 		// Ignore ARP replies
-		if r.Operation != arp.OperationRequest {
-			return
+		if pkt.Operation != arp.OperationRequest {
+			continue
 		}
 
 		// Ignore ARP requests which are not broadcast or bound directly for
 		// this machine
-		if !bytes.Equal(r.TargetHardwareAddr, ethernet.Broadcast) && !bytes.Equal(r.TargetHardwareAddr, ifi.HardwareAddr) {
-			return
+		if !bytes.Equal(pkt.TargetHardwareAddr, ethernet.Broadcast) && !bytes.Equal(pkt.TargetHardwareAddr, ifi.HardwareAddr) {
```

Comment:
Do note that this is wrong. The ethernet destination should be the broadcast address, yes. but the target hardware address should be all zeroes. The second check is likely to be wrong, too, and should also refer to the ethernet header instead.

Will fix this in a separate PR, though, to keep this one clean from unrelated changes.